core/commands/dag/dag.go

```diff
@@ -241,3 +278,421 @@ var DagResolveCmd = &cmds.Command{
+				ret <- importResult{err: err}
+				return
+			}
+
+			counts.Blocks++
+
+			// encountered something known to be a root, for the first time
+			if seen, exists := expectedRootsSeen[nd.Cid()]; exists && !seen {
+				expectedRootsSeen[nd.Cid()] = true
+				counts.RootsSeen++
+			}
+		}
+	}
+
+	if err := it.Err(); err != nil {
+		ret <- importResult{err: err}
+		return
+	}
+
+	if err := batch.Commit(); err != nil {
+		ret <- importResult{err: err}
+		return
+	}
+
+	ret <- importResult{roots: &expectedRootsSeen}
+}
+
+var DagExportCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Streams the selected DAG as a .car stream on stdout.",
+		ShortDescription: `
+'ipfs dag export' fetches a dag and streams it out as a well-formed .car file.
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("selector", true, false, "Selector representing the dag to export").EnableStdin(),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption(progressOptionName, "p", "Display progress on CLI.").WithDefault(true),
+	},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+
+		c, err := cid.Decode(req.Arguments[0])
+		if err != nil {
+			return fmt.Errorf(
+				"unable to parse selector (currently only bare CIDs are supported): %s",
+				err,
+			)
+		}
+
+		// The current interface of go-car is rather suboptimal as it
+		// only takes a blockstore, instead of accepting a dagservice,
+		// and leveraging parallel-fetch capabilities
+		//
+		// Until the above is fixed, pre-warm the blockstore before doing
+		// anything else. We explicitly *DO NOT* take a lock during this
+		// operation: even if we lose some of the blocks we just received
+		// due to a conflicting GC: we will just re-retrieve everything when
+		// the car is being streamed out
+		node, err := cmdenv.GetNode(env)
+		if err != nil {
+			return err
+		}
+
+		if err := mdag.FetchGraph(req.Context, c, node.DAG); err != nil {
+			if !node.IsOnline {
+				err = fmt.Errorf("%s (currently offline, perhaps retry after attaching to the network)", err)
+			}
+			return err
+		}
+
+		sess := blockservice.NewSession(
+			req.Context,
+			node.Blocks,
+		)
+
+		var nodeCount int64
+
+		// The second part of the above - make a super-thin wrapper around
+		// a blockservice session, translating Session.GetBlock() to Blockstore.Get()
+		// The function also doubles as our progress meter
+		var wrapper getBlockFromSessionWrapper = func(c cid.Cid) (blk.Block, error) {
+			nodeCount++
+			return sess.GetBlock(req.Context, c)
+		}
+		sb := gipselectorbuilder.NewSelectorSpecBuilder(gipfree.NodeBuilder())
+		car := gocar.NewSelectiveCar(
+			req.Context,
+			&wrapper,
+			[]gocar.Dag{gocar.Dag{
+				Root: c,
+				Selector: sb.ExploreRecursive(
+					gipselector.RecursionLimitNone(),
+					sb.ExploreAll(sb.ExploreRecursiveEdge()),
+				).Node(),
+			}},
+		)
+
+		if showProgress, _ := req.Options[progressOptionName].(bool); showProgress {
+			stopProgress := make(chan bool)
```

Comment:
rather than the channel, might be more conventional at this point to make a `context.WithCancel`

Reply:
Ugh... this is ( still ) above my head. I suspect it will be faster for me to grok how to do this in the future if you just push a commit directly addressing this...?